plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java

```diff
@@ -2239,211 +2305,309 @@ public LibvirtVMDef createVMFromSpec(final VirtualMachineTO vmTO) {
-            }
+    /**
+     * Add the VNC port passwd here, get the passwd from the vmInstance.
+     */
+    protected GraphicDef createGraphicDef(VirtualMachineTO vmTO) {
+        return new GraphicDef(VNC, (short)0, true, vmTO.getVncAddr(), vmTO.getVncPassword(), null);
+    }
 
-            setQuotaAndPeriod(vmTO, ctd);
+    /**
+     * Add a VirtIO channel for the Qemu Guest Agent tools.
+     */
+    protected ChannelDef createChannelDef(VirtualMachineTO vmTO) {
+        File virtIoChannel = Paths.get(_qemuSocketsPath.getPath(), vmTO.getName() + "." + _qemuGuestAgentSocketName).toFile();
+        return new ChannelDef(_qemuGuestAgentSocketName, ChannelDef.ChannelType.UNIX, virtIoChannel);
+    }
 
-            vm.addComp(ctd);
-        }
+    /**
+     * If we're using virtio scsi, then we need to add a virtual scsi controller.
+     */
+    protected SCSIDef createSCSIDef(int vcpus) {
+        return new SCSIDef((short)0, 0, 0, 9, 0, vcpus);
+    }
 
-        final FeaturesDef features = new FeaturesDef();
-        features.addFeatures("pae");
-        features.addFeatures("apic");
-        features.addFeatures("acpi");
-        if (isUefiEnabled && isSecureMode(customParams.get(GuestDef.BootType.UEFI.toString()))) {
-            features.addFeatures("smm");
-        }
+    protected ConsoleDef createConsoleDef() {
+        return new ConsoleDef(PTY, null, null, (short)0);
+    }
 
-        //KVM hyperv enlightenment features based on OS Type
-        enlightenWindowsVm(vmTO, features);
+    protected VideoDef createVideoDef() {
+        return new VideoDef(_videoHw, _videoRam);
+    }
 
-        vm.addComp(features);
+    protected RngDef createRngDef() {
+        return new RngDef(_rngPath, _rngBackendModel, _rngRateBytes, _rngRatePeriod);
+    }
 
-        final TermPolicy term = new TermPolicy();
-        term.setCrashPolicy("destroy");
-        term.setPowerOffPolicy("destroy");
-        term.setRebootPolicy("restart");
-        vm.addComp(term);
+    protected SerialDef createSerialDef() {
+        return new SerialDef(PTY, null, (short)0);
+    }
 
-        final ClockDef clock = new ClockDef();
-        if (vmTO.getOs().startsWith("Windows")) {
+    protected ClockDef createClockDef(final VirtualMachineTO vmTO) {
+        ClockDef clock = new ClockDef();
+        if (org.apache.commons.lang.StringUtils.startsWith(vmTO.getOs(), WINDOWS)) {
             clock.setClockOffset(ClockDef.ClockOffset.LOCALTIME);
-            clock.setTimer("hypervclock", null, null);
-        } else if (vmTO.getType() != VirtualMachine.Type.User || isGuestPVEnabled(vmTO.getOs())) {
-            if (_hypervisorLibvirtVersion >= 9 * 1000 + 10) {
-                clock.setTimer("kvmclock", null, null, _noKvmClock);
-            }
+            clock.setTimer(HYPERVCLOCK, null, null);
+        } else if ((vmTO.getType() != VirtualMachine.Type.User || isGuestPVEnabled(vmTO.getOs())) && _hypervisorLibvirtVersion >= MIN_LIBVIRT_VERSION_FOR_GUEST_CPU_MODE) {
```

Comment:
This if is quite big, maybe extract it to a method?

Reply:
I can't understand what would be the use of extracting this if for a method, considering that it is not used anywhere else. Could you explain your point?

Reply:
That's a good question @SadiJr. In fact, some code extractions have little to do with duplicated code but with making cleaner code and also helping with unit test coverage.

For instance, there are cases where conditionals can be replaced by polymorphism. In some cases, it is also possible to replace a conditional with map/set. As an example:
```
if (var == value1 || var == value2 || ... var == valueN) {
    <body>
}
```
:arrow_down: 
```
/**
 * This JavaDoc might help. But if the variable name is self-explanatory this Javadoc might be unnecessary.
 */
final static Set<Type> GOOD_VAR_NAME = new HashSet<>(Arrays.asList(value1, value2, ..., valueN));
....
protected void method() {
   ....
   if (GOOD_VAR_NAME.contains(var)) {
       <body>
   }
}
```

Therefore, the goal of such extraction would be to:
1. keep code clean
2. add documentation (or not)
3. allow easy JUnit test that covers such if-conditional

You can proceed with the code as it is or look for some alternatives that might improve this IF. I would not block or judge if you keep it as it is. What matters the most is to always try to keep such care when developing/reviewing.

Reply:
Now I get your point. However, while I don't see any problem making this change, I also don't see the need to make it, so I'll keep the code as is. But thanks for the clarification, it was a great explanation.